File: helpers/package-json-scripts.py
#!/usr/bin/env python3
import json
import os
import shutil
import subprocess
import sys
from pathlib import Path

PLUGIN_ID = "package"
PLUGIN_TITLE = "Package Scripts"


def find_package_json(base: Path) -> Path | None:
    current = base
    for candidate in [current, *current.parents]:
        pkg = candidate / "package.json"
        if pkg.is_file():
            return pkg
    return None


def load_scripts(pkg_path: Path) -> dict[str, str]:
    data = json.loads(pkg_path.read_text(encoding="utf-8"))
    scripts = data.get("scripts", {})
    if not isinstance(scripts, dict):
        return {}
    out: dict[str, str] = {}
    for name, command in scripts.items():
        if isinstance(name, str) and name and isinstance(command, str):
            out[name] = command
    return out


def describe() -> int:
    cwd = Path.cwd()
    pkg_path = find_package_json(cwd)
    tasks = []

    if pkg_path:
        scripts = load_scripts(pkg_path)
        for name, command in sorted(scripts.items()):
            tasks.append(
                {
                    "name": name,
                    "title": f"Run {name}",
                    "group": "Node",
                    "description": command,
                    "inputs": [],
                }
            )

    manifest = {
        "schemaVersion": 1,
        "plugin": {
            "id": PLUGIN_ID,
            "title": PLUGIN_TITLE,
            "version": "0.1.0",
        },
        "tasks": tasks,
    }
    print(json.dumps(manifest))
    return 0


def pick_runner(workdir: Path) -> list[str]:
    if (workdir / "pnpm-lock.yaml").is_file() and shutil.which("pnpm"):
        return ["pnpm", "run"]
    if (workdir / "yarn.lock").is_file() and shutil.which("yarn"):
        return ["yarn", "run"]
    if shutil.which("npm"):
        return ["npm", "run"]
    if shutil.which("yarn"):
        return ["yarn", "run"]
    if shutil.which("pnpm"):
        return ["pnpm", "run"]
    print("no package manager found (npm/yarn/pnpm)", file=sys.stderr)
    sys.exit(1)


def read_payload() -> dict:
    raw = sys.stdin.read().strip()
    if not raw:
        return {}
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        print(f"invalid input JSON: {exc}", file=sys.stderr)
        sys.exit(1)
    if not isinstance(data, dict):
        print("input JSON must be an object", file=sys.stderr)
        sys.exit(1)
    return data


def run(task_name: str) -> int:
    payload = read_payload()
    ctx = payload.get("ctx", {})
    if not isinstance(ctx, dict):
        ctx = {}

    cwd = ctx.get("cwd") or os.getenv("AUTOMATE_ME_CWD") or os.getcwd()
    workdir = Path(cwd)
    pkg_path = find_package_json(workdir)
    if not pkg_path:
        print("package.json not found from current context", file=sys.stderr)
        return 1

    scripts = load_scripts(pkg_path)
    if task_name not in scripts:
        print(f"unknown script: {task_name}", file=sys.stderr)
        return 1

    runner = pick_runner(pkg_path.parent)
    cmd = [*runner, task_name]
    proc = subprocess.run(cmd, cwd=str(pkg_path.parent))
    return proc.returncode


def usage() -> int:
    print("usage: package-json-scripts describe|run <task>", file=sys.stderr)
    return 1


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        return usage()
    cmd = argv[1]
    if cmd == "describe":
        return describe()
    if cmd == "run":
        if len(argv) < 3:
            return usage()
        return run(argv[2])
    return usage()


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: helpers/package-json-scripts.js
#!/usr/bin/env node

const fs = require('node:fs');
const path = require('node:path');
const { spawnSync } = require('node:child_process');

const PLUGIN_ID = 'package';
const PLUGIN_TITLE = 'Package Scripts';

function findPackageJson(startDir) {
  let current = path.resolve(startDir);
  while (true) {
    const pkg = path.join(current, 'package.json');
    if (fs.existsSync(pkg) && fs.statSync(pkg).isFile()) {
      return pkg;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      return null;
    }
    current = parent;
  }
}

function loadScripts(pkgPath) {
  let data;
  try {
    data = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
  } catch (err) {
    console.error(`invalid package.json: ${err.message}`);
    process.exit(1);
  }

  if (!data || typeof data !== 'object' || !data.scripts || typeof data.scripts !== 'object') {
    return {};
  }

  const scripts = {};
  for (const [name, cmd] of Object.entries(data.scripts)) {
    if (typeof name === 'string' && name && typeof cmd === 'string') {
      scripts[name] = cmd;
    }
  }
  return scripts;
}

function describe() {
  const pkgPath = findPackageJson(process.cwd());
  const tasks = [];

  if (pkgPath) {
    const scripts = loadScripts(pkgPath);
    for (const name of Object.keys(scripts).sort()) {
      tasks.push({
        name,
        title: `Run ${name}`,
        group: 'Node',
        description: scripts[name],
        inputs: [],
      });
    }
  }

  const manifest = {
    schemaVersion: 1,
    plugin: {
      id: PLUGIN_ID,
      title: PLUGIN_TITLE,
      version: '0.1.0',
    },
    tasks,
  };

  process.stdout.write(`${JSON.stringify(manifest)}\n`);
}

function hasCommand(cmd) {
  const checker = process.platform === 'win32' ? 'where' : 'command';
  const args = process.platform === 'win32' ? [cmd] : ['-v', cmd];
  const res = spawnSync(checker, args, { stdio: 'ignore', shell: process.platform !== 'win32' });
  return res.status === 0;
}

function pickRunner(workdir) {
  if (fs.existsSync(path.join(workdir, 'pnpm-lock.yaml')) && hasCommand('pnpm')) {
    return ['pnpm', 'run'];
  }
  if (fs.existsSync(path.join(workdir, 'yarn.lock')) && hasCommand('yarn')) {
    return ['yarn', 'run'];
  }
  if (hasCommand('npm')) {
    return ['npm', 'run'];
  }
  if (hasCommand('yarn')) {
    return ['yarn', 'run'];
  }
  if (hasCommand('pnpm')) {
    return ['pnpm', 'run'];
  }
  console.error('no package manager found (npm/yarn/pnpm)');
  process.exit(1);
}

function readPayload() {
  const raw = fs.readFileSync(0, 'utf8').trim();
  if (!raw) {
    return {};
  }
  try {
    const data = JSON.parse(raw);
    if (data && typeof data === 'object') {
      return data;
    }
    console.error('input JSON must be an object');
    process.exit(1);
  } catch (err) {
    console.error(`invalid input JSON: ${err.message}`);
    process.exit(1);
  }
}

function run(taskName) {
  const payload = readPayload();
  const ctx = payload && typeof payload.ctx === 'object' ? payload.ctx : {};
  const cwd = ctx.cwd || process.env.AUTOMATE_ME_CWD || process.cwd();

  const pkgPath = findPackageJson(cwd);
  if (!pkgPath) {
    console.error('package.json not found from current context');
    process.exit(1);
  }

  const scripts = loadScripts(pkgPath);
  if (!Object.prototype.hasOwnProperty.call(scripts, taskName)) {
    console.error(`unknown script: ${taskName}`);
    process.exit(1);
  }

  const workdir = path.dirname(pkgPath);
  const runner = pickRunner(workdir);
  const result = spawnSync(runner[0], [...runner.slice(1), taskName], {
    cwd: workdir,
    stdio: 'inherit',
  });

  if (result.error) {
    console.error(result.error.message);
    process.exit(1);
  }
  process.exit(result.status || 0);
}

function usage() {
  console.error('usage: package-json-scripts.js describe|run <task>');
  process.exit(1);
}

function main(argv) {
  if (argv.length < 3) {
    usage();
  }
  const cmd = argv[2];
  if (cmd === 'describe') {
    describe();
    return;
  }
  if (cmd === 'run') {
    if (argv.length < 4) {
      usage();
    }
    run(argv[3]);
    return;
  }
  usage();
}

main(process.argv);
